crypto/rand/rand_lib.c

```diff
@@ -36,15 +112,51 @@
 # include "internal/e_os.h"
 # include "internal/property.h"
 
+/*
+ * The default name for the random provider.
+ * This ensures that the FIPS provider will supply libcrypto's random byte
+ * requirements.
+ */
+static const char random_provider_fips_name[] = "fips";
+
+/*
+ * Utility functions to help handling the default random provider name.
+ * A pointer compare is not sufficient to detect a name being unmodified
+ * because a loaded provider can have its own copy of the global on some
+ * platforms (MIPS).  We therefore resort to string comparisons.
+ */
+static void free_random_provider_name(RAND_GLOBAL *dgbl)
+{
+    if (dgbl->random_provider_name != NULL
+            && strcmp(dgbl->random_provider_name,
+                      random_provider_fips_name) != 0)
+        OPENSSL_free((void *)dgbl->random_provider_name);
+}
```

Comment:
This code should also set dgbl->random_provider_name to NULL after free - we shouldn't leave it around with a now invalid value.